app/src/main/kotlin/com/wire/android/services/ServicesManager.kt

```diff
@@ -25,26 +25,109 @@ import android.content.Context
+                                // happens before the service calls startForeground, we call the startService command with an empty data
+                                // or some specific argument that tells the service that it should stop itself right after startForeground.
+                                // This way, when this service is killed and recreated by the system, it will stop itself right after
+                                // recreating so it won't cause any problems.
+                                startService(OngoingCallService.newIntentToStop(context))
+                                appLogger.i("ServicesManager: OngoingCallService stopped by passing stop argument")
+                            }
+                            OngoingCallService.ServiceState.FOREGROUND -> {
+                                // we can just stop the service, because it's already in foreground
+                                context.stopService(OngoingCallService.newIntent(context))
+                                appLogger.i("ServicesManager: OngoingCallService stopped by calling stopService")
+                            }
+                            else -> {
+                                appLogger.i("ServicesManager: OngoingCallService not running, nothing to stop")
+                            }
+                        }
+                    } else {
+                        appLogger.i("ServicesManager: starting OngoingCallService")
+                        startService(OngoingCallService.newIntent(context))
+                    }
+                }
+        }
+    }
 
     // Ongoing call
-    fun startOngoingCallService(notificationTitle: String, conversationId: ConversationId, userId: UserId) {
-        val onGoingCallService = OngoingCallService.newIntent(context, userId.toString(), conversationId.toString(), notificationTitle)
-        startService(onGoingCallService)
+    fun startOngoingCallService(ongoingCallData: OngoingCallData) {
+        scope.launch {
+            appLogger.i("ServicesManager: starting OngoingCallService for user:${ongoingCallData.userId.toLogString()}")
+            ongoingCallServiceForUsers.update { it.filter { it.userId != ongoingCallData.userId } + ongoingCallData }
+        }
     }
 
-    fun stopOngoingCallService() {
-        stopService(OngoingCallService::class)
+    fun stopOngoingCallServiceForUser(userId: UserId) {
+        scope.launch {
+            appLogger.i("ServicesManager: stopping OngoingCallService for user:${userId.toLogString()}")
+            ongoingCallServiceForUsers.update { it.filter { it.userId != userId } }
+        }
+    }
+
+    fun stopOngoingCallServiceForAll() {
+        scope.launch {
+            appLogger.i("ServicesManager: stopping OngoingCallService for all users")
+            ongoingCallServiceForUsers.emit(emptyList())
+        }
```

Comment:
I know, but checks for each account are done independently, so if we have users A and B, user A has an ongoing call and B does not, if the order in which app gets this info is first A then B, then the service will start and stop immediately, so this solution is to provide a buffer to gather data from all users instead of relying on only one of them - now if the app gets the info that user A needs a service, if then user B comes up and tells that doesn't need it, the service will still be running because we still keep the info that user A needs it.
"stop for user" means that we let the app know that this user doesn't need this service, "stop for all" means that we should mark all users as not needing the service and stop the service instantly.

Names of these functions are not the best, so I changed them to unify it with handling incoming calls (one method "handle" instead of "start" and "stop for user"), let me know if that's ok or maybe do you have any ideas for better names for them?

Reply:
I got your point, but what's the issue if the service and stop immediately ? if it's a crash, the delay that we added shouldn't fix it

Reply:
The crash happens because `stopService` happens too soon, before `startForeground` is executed, so adding a debounce of 200millis is usually enough for the app to execute it, and even if not, then I also implemented a mechanism to do it in right order, so maybe debounce is redundant but it will actually also reduce starting and stopping the service if not needed so the app won't use resources when not needed